vision/convmixer_cifar10/convmixer.jl

```diff
@@ -1,139 +1,143 @@
-            l += Flux.logitcrossentropy(z, y, agg=sum)
-            acc += sum(onecold(z).==onecold(y))
-            n += size(x)[end]
-        end
-        l / n, acc / n
+"""
+This function is used only for logging, with either train_loader or test_loader.
+"""
+function loss_and_accuracy(model, dataloader)
+    n = 0      # count images seen
+    l = 0.0f0  # total loss
+    c = 0      # count correct classifications
+
+    for (x1, y1) in dataloader
+        x, y = gpu(x1), gpu(y1)  # assuming model is on gpu too!
+        z = model(x)
+        l += Flux.logitcrossentropy(z, y, agg=sum)
+        c += sum(Flux.onecold(z) .== Flux.onecold(y))
+        n += size(x)[end]
     end
-
-    return loss
-   
+    return l/n, c/n  # mean loss, and accuracy
 end
 
+"""
+Main function: loads data, creates model, trains it, then saves trained model in "model.bson"
+Keywords `train(epochs=5, images=99)` will use only the first 99 images of the training set
+(to check that it runs).
+"""
+function train(; epochs=100, images=:)
+    train_loader, test_loader = get_data(128; images)
 
-function train(n_epochs=100)
-
-    #params: warning, the training can be long with these params
-    train_loader, test_loader = get_data(128)
-    η = 3e-4
+    # hyper-parameters. Note that training can be long with these!
     in_channel = 3
     patch_size = 2
     kernel_size = 7
     dim = 128
-    dimPL = 2
     depth = 18
-    use_cuda = true
 
-    #logging the losses
-    train_save = zeros(n_epochs, 2)
-    test_save = zeros(n_epochs, 2)
+    model = make_convmixer(in_channel, kernel_size, patch_size, dim, depth, 10) |> gpu
 
-    if use_cuda
-        device = gpu
-        @info "Training on GPU"
-    else
-        device = cpu
-        @info "Training on CPU"
-    end
-
-    train_loss_fn = create_loss_function(train_loader, device)
-    test_loss_fn = create_loss_function(test_loader, device)
-
-    model = ConvMixer(in_channel, kernel_size, patch_size, dim, depth, 10) |> device
-
-    ps = params(model)
-    opt = Optimiser(
-            WeightDecay(1f-3), 
-            ClipNorm(1.0),
-            ADAM(η)
+    # optimiser and its state:
+    opt = OptimiserChain(
+            WeightDecay(1f-3),  # L2 regularisation
+            ClipNorm(1f0),
+            Adam(3f-4),  # learning rate
             )
-
-    for epoch in 1:n_epochs
-        for (x,y) in train_loader
-            x,y = x|>device, y|>device
-            gr = gradient(()->Flux.logitcrossentropy(model(x), y, agg=sum), ps)
-            Flux.Optimise.update!(opt, ps, gr)
+    state = Flux.setup(opt, model)
+
+    # arrays for logging:
+    train_save = zeros(epochs, 2)
+    test_save = zeros(epochs, 2)
+
+    for epoch in 1:epochs
+        for (x1, y1) in train_loader
+            # move one batch at a time to GPU; gpu(train_loader) would be another way
+            x, y = gpu(x1), gpu(y1)
+            grads = gradient(m -> Flux.logitcrossentropy(m(x), y; agg=sum), model)
```

Comment:
why `agg=sum`?

Reply:
I just left this, as it probably doesn't hurt... no idea why it's there, maybe it was like that in what someone followed? Showing what the keyword is doesn't seem awful